src/ORM/Related/DataObjectRelatedService.php

```diff
@@ -0,0 +1,332 @@
+                    // $componentClass will usually be a File, so a $record subclass like MyFile will match
+                    if (!($record instanceof $componentClass)) {
+                        continue;
+                    }
+
+                    // Query database to get file usage data
+                    $results = [];
+                    if ($relationType === 'has_one') {
+                        // Ensure table exists, this is required for TestOnly SapphireTest classes
+                        if (!isset($classToTableName[$class])) {
+                            continue;
+                        }
+
+                        // Example: class MyPage ... private static $has_one = [ 'MyFile' => File::class ]
+                        $componentIDField = "{$component}ID";
+
+                        // Only get database fields from the current class model, not parent class model
+                        $dbFields = $dataObjectSchema->databaseFields($class, false);
+                        if (!isset($dbFields[$componentIDField])) {
+                            continue;
+                        }
+                        $tableName = $dataObjectSchema->tableName($class);
+
+                        // Example SQL: SELECT "ID" FROM "MyPage" WHERE "MyFileID" = 123;
+                        $results = SQLSelect::create(
+                            '"ID"',
+                            sprintf('"%s"', $tableName),
+                            sprintf('"%s" = %s', $componentIDField, $record->ID)
+                        )->execute();
+                    }
+                    if ($relationType === 'many_many' || $relationType === 'belongs_many_many') {
+                        $results = $this->fetchManyManyResults(
+                            $record,
+                            $class,
+                            $component,
+                            false,
+                            $dataObjectSchema,
+                            $classToTableName
+                        );
+                    }
+                    $this->updateClassIDs($classIDs, $results, $class);
+                }
+            }
+        }
+
+        // Loop has_one relationships on the DataObject we're getting usage for
+        // e.g. File.has_one = Page, Page.has_many = File
+        $class = get_class($record);
+        foreach (Config::forClass($class)->get('has_one') as $component => $componentClass) {
+            $componentClass = $this->updateComponentClass($componentClass, $config, $throughClasses);
+            $componentIDField = "{$component}ID";
+            // Find the table that contains $componentIDField - this is relevant for subclassed DataObjects
+            $tableName = '';
+            $candidateClass = $class;
+            // Using a for loop rather than a while loop just to be safe
+            for ($i = 0; $i < 25 && $candidateClass; $i++) {
```

Comment:
Are we expecting some sort of recursion here? We ought to have a good understanding of our exit condition here and make sure it's always met at some stage rather than rely on an arbitrary loop limit.

Reply:
Changed it to a while loop